api_app/services/aad_authentication.py

```diff
@@ -317,6 +358,143 @@ def get_workspace_user_emails_by_role_assignment(self, workspace: Workspace):
+                            isEnabled=role["isEnabled"],
+                            description=role["description"],
+                            displayName=role["displayName"],
+                            origin=role["origin"],
+                            allowedMemberTypes=role["allowedMemberTypes"])
+
+        return None
+
+    def get_user_by_email(self, user_email: str) -> User:
+        encoded_email = urllib.parse.quote(user_email)
+        user_by_email_endpoint = f"{MICROSOFT_GRAPH_URL}/v1.0/users/{encoded_email}"
+        user_data = self._ms_graph_query(user_by_email_endpoint, "GET")
+
+        return User(id=user_data["id"][0], name=user_data["displayName"][0], email=user_email, roles=[])
+
+    def assign_workspace_user(self, user: User, workspace: Workspace, role: Role) -> None:
+        # User already has the role, do nothing
+        if self._is_user_in_role(user, role):
+            return
+        if self._is_workspace_role_group_in_use(workspace):
+            return self._assign_workspace_user_to_application_group(user, workspace, role)
+        else:
+            return self._assign_workspace_user_to_application(user, workspace, role)
+
+    def _is_user_in_role(self, user: User, role: Role) -> bool:
+        return any(r for r in user.roles if r == role.value)
+
+    def _is_workspace_role_group_in_use(self, workspace: Workspace) -> bool:
+        aad_groups_in_user = workspace.properties["create_aad_groups"]
+        return aad_groups_in_user
+
+    def _get_workspace_group_name(self, workspace: Workspace, role: Role) -> tuple:
+        tre_id = workspace.properties["tre_id"]
+        workspace_id = workspace.properties["workspace_id"]
+        group_suffix = ""
+        app_role_id_suffix = ""
+        if role.value == "WorkspaceResearcher":
+            group_name = "Workspace Researchers"
+            app_role_id_suffix = "workspace_researcher"
+        elif role.value == "WorkspaceOwner":
+            group_name = "Workspace Owners"
+            app_role_id_suffix = "workspace_owner"
+        elif role.value == "AirlockManager":
+            group_name = "Airlock Managers"
+            app_role_id_suffix = "workspace_airlock_manager"
+        else:
+            raise UserRoleAssignmentError(f"Unknown role: {role.value}")
+
+        return (f"{tre_id}-ws-{workspace_id} {group_name}", f"app_role_id_{app_role_id_suffix}")
+
+    def _assign_workspace_user_to_application_group(self, user: User, workspace: Workspace, role: Role):
+        msgraph_token = self._get_msgraph_token()
+        roles_graph_data = self._get_user_role_assignments(workspace.properties["sp_id"], msgraph_token)
+        group_details = self._get_workspace_group_name(workspace, role)
+        group_name = group_details[0]
+        workspace_app_role_field = group_details[1]
+
+        for group in [item for item in roles_graph_data["value"] if item["principalType"] == PrincipalType.Group.value]:
```

Comment:
Bit confused, might need to talk through it. How does roles_graph_data["value"] contain groups? looks to be getting existing user role assignments?

Reply:
roles_graph_data contains a list of appRoles assigned to the API, it is a mixture of users and groups

Reply:
Ah, got it, then look through those groups by name?

If it works, ok, think using the workspace outputs would be more straightforward.